File: frontend/src/modeling/lib/mesh-arrays.ts
export class MeshArrays {
    vertices: Float32Array;
    normals: Float32Array;
    uvs: Float32Array;
    ao: Float32Array;
    isSelected: Float32Array;
    indices: Uint32Array;
    vertexCount: number = 0;
    indexCount: number = 0;

    constructor(maxVertices: number, maxIndices: number) {
        this.vertices = new Float32Array(maxVertices * 3);
        this.normals = new Float32Array(maxVertices * 3);
        this.uvs = new Float32Array(maxVertices * 2);
        this.ao = new Float32Array(maxVertices);
        this.isSelected = new Float32Array(maxVertices);
        this.indices = new Uint32Array(maxIndices);
    }

    reset() {
        this.vertexCount = 0;
        this.indexCount = 0;
    }

    pushVertex(x: number, y: number, z: number): void {
        const offset = this.vertexCount * 3;
        this.vertices[offset] = x;
        this.vertices[offset + 1] = y;
        this.vertices[offset + 2] = z;
    }

    pushNormal(x: number, y: number, z: number): void {
        const offset = this.vertexCount * 3;
        this.normals[offset] = x;
        this.normals[offset + 1] = y;
        this.normals[offset + 2] = z;
    }

    pushUV(u: number, v: number): void {
        const offset = this.vertexCount * 2;
        this.uvs[offset] = u;
        this.uvs[offset + 1] = v;
    }

    pushAO(value: number): void {
        this.ao[this.vertexCount] = value;
    }

    pushIsSelected(value: number): void {
        this.isSelected[this.vertexCount] = value;
    }

    incrementVertex(): number {
        return this.vertexCount++;
    }

    pushIndex(index: number): void {
        this.indices[this.indexCount++] = index;
    }

    getVertices() {
        return this.vertices.subarray(0, this.vertexCount * 3);
    }

    getNormals() {
        return this.normals.subarray(0, this.vertexCount * 3);
    }

    getUVs() {
        return this.uvs.subarray(0, this.vertexCount * 2);
    }

    getAO() {
        return this.ao.subarray(0, this.vertexCount);
    }

    getIsSelected() {
        return this.isSelected.subarray(0, this.vertexCount);
    }

    getIndices() {
        return this.indices.subarray(0, this.indexCount);
    }
}

File: wasm/src/mesh_arrays.rs
pub struct MeshArrays {
    pub vertices: Vec<f32>,
    pub normals: Vec<f32>,
    pub uvs: Vec<f32>,
    pub ao: Vec<f32>,
    pub is_selected: Vec<f32>,
    pub indices: Vec<u32>,
    pub vertex_count: usize,
    pub index_count: usize,
}

impl MeshArrays {
    pub fn new(max_vertices: usize, max_indices: usize) -> Self {
        Self {
            vertices: vec![0.0; max_vertices * 3],
            normals: vec![0.0; max_vertices * 3],
            uvs: vec![0.0; max_vertices * 2],
            ao: vec![0.0; max_vertices],
            is_selected: vec![0.0; max_vertices],
            indices: vec![0; max_indices],
            vertex_count: 0,
            index_count: 0,
        }
    }

    pub fn reset(&mut self) {
        self.vertex_count = 0;
        self.index_count = 0;
    }

    #[inline(always)]
    pub fn push_vertex(&mut self, x: f32, y: f32, z: f32) {
        let offset = self.vertex_count * 3;
        self.vertices[offset] = x;
        self.vertices[offset + 1] = y;
        self.vertices[offset + 2] = z;
    }

    #[inline(always)]
    pub fn push_normal(&mut self, x: f32, y: f32, z: f32) {
        let offset = self.vertex_count * 3;
        self.normals[offset] = x;
        self.normals[offset + 1] = y;
        self.normals[offset + 2] = z;
    }

    #[inline(always)]
    pub fn push_uv(&mut self, u: f32, v: f32) {
        let offset = self.vertex_count * 2;
        self.uvs[offset] = u;
        self.uvs[offset + 1] = v;
    }

    #[inline(always)]
    pub fn push_ao(&mut self, value: f32) {
        self.ao[self.vertex_count] = value;
    }

    #[inline(always)]
    pub fn push_is_selected(&mut self, value: u8) {
        self.is_selected[self.vertex_count] = value as f32;
    }

    #[inline(always)]
    pub fn increment_vertex(&mut self) {
        self.vertex_count += 1;
    }

    #[inline(always)]
    pub fn push_index(&mut self, index: u32) {
        self.indices[self.index_count] = index;
        self.index_count += 1;
    }
}
